File: app/src/main/java/com/senierr/demo/MainActivity.java
package com.senierr.demo;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import com.senierr.adapter.internal.DataBinder;
import com.senierr.adapter.internal.MultiTypeAdapter;
import com.senierr.adapter.internal.ViewHolder;
import com.senierr.adapter.listener.OnChildClickListener;
import com.senierr.adapter.listener.OnChildLongClickListener;
import com.senierr.adapter.listener.OnItemClickListener;
import com.senierr.adapter.listener.OnItemLongClickListener;
import com.senierr.adapter.support.bean.LoadMoreBean;
import com.senierr.adapter.support.bean.StateBean;
import com.senierr.adapter.support.wrapper.BaseLoadMoreWrapper;
import com.senierr.demo.wrapper.FirstWrapper;
import com.senierr.demo.wrapper.LoadMoreWrapper;
import com.senierr.demo.wrapper.SecondWrapper;
import com.senierr.demo.wrapper.StateWrapper;

import java.util.List;

public class MainActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private MultiTypeAdapter multiTypeAdapter;

    private LoadMoreWrapper loadMoreWrapper;
    private StateWrapper stateWrapper;

    private int pageIndex = 1;
    private int pageSize = 20;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initView();
        loadData();
    }

    /**
     * 初始化界面
     */
    private void initView() {
        recyclerView = findViewById(R.id.recyclerView);
        recyclerView.setHasFixedSize(true);
//        recyclerView.setLayoutManager(new LinearLayoutManager(this));
//        recyclerView.setLayoutManager(new GridLayoutManager(this, 3));
        recyclerView.setLayoutManager(new StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL));
        multiTypeAdapter = new MultiTypeAdapter();

        FirstWrapper firstWrapper = new FirstWrapper();
        // 列表点击事件
        firstWrapper.setOnItemClickListener(new OnItemClickListener<DataBean>() {
            @Override
            public void onClick(ViewHolder viewHolder, int position, DataBean dataBean) {
                showToast("ItemClick: " + dataBean.getContent());
            }
        });
        firstWrapper.setOnItemLongClickListener(new OnItemLongClickListener<DataBean>() {
            @Override
            public boolean onLongClick(ViewHolder viewHolder, int position, DataBean dataBean) {
                showToast("ItemLongClick: " + dataBean.getContent());
                return true;
            }
        });
        // 子控件点击事件
        firstWrapper.setOnChildClickListener(R.id.btn_click, new OnChildClickListener<DataBean>() {
            @Override
            public void onClick(ViewHolder viewHolder, View view, int position, DataBean dataBean) {
                showToast("ChildClick: " + dataBean.getContent());
            }
        });
        firstWrapper.setOnChildLongClickListener(R.id.btn_click, new OnChildLongClickListener<DataBean>() {
            @Override
            public boolean onLongClick(ViewHolder viewHolder, View view, int position, DataBean dataBean) {
                showToast("ChildLongClick: " + dataBean.getContent());
                return true;
            }
        });
        SecondWrapper secondWrapper = new SecondWrapper();

        // 加载更多
        loadMoreWrapper = new LoadMoreWrapper();
        loadMoreWrapper.setOnLoadMoreListener(new BaseLoadMoreWrapper.OnLoadMoreListener() {
            @Override
            public void onLoadMore() {
                loadData();
            }
        });
        // 状态切换
        stateWrapper = new StateWrapper();
        stateWrapper.setOnItemClickListener(new OnItemClickListener<StateBean>() {
            @Override
            public void onClick(ViewHolder viewHolder, int position, StateBean stateBean) {
                pageIndex = 1;
                loadData();
            }
        });

        multiTypeAdapter.bind(DataBean.class, firstWrapper, secondWrapper)
                .with(new DataBinder<DataBean>() {
                    @Override
                    public int onBindIndex(@NonNull DataBean item) {
                        if (item.getId() % 2 == 0) {
                            return 0;
                        }
                        return 1;
                    }
                });
        multiTypeAdapter.bind(StateBean.class, stateWrapper);
        multiTypeAdapter.bind(LoadMoreBean.class, loadMoreWrapper);
        recyclerView.setAdapter(multiTypeAdapter);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_default) {
            stateWrapper.setState(StateWrapper.STATE_DEFAULT);
        }
        return super.onOptionsItemSelected(item);
    }

    /**
     * 加载数据
     */
    private void loadData() {
        recyclerView.postDelayed(new Runnable() {
            @Override
            public void run() {
                List<DataBean> dataBeanList = DataBean.getData(pageIndex, pageSize);
                if (pageIndex == 1) {
                    multiTypeAdapter.clearData();
                    multiTypeAdapter.addDatas(dataBeanList);
                    multiTypeAdapter.addData(loadMoreWrapper.getLoadMoreBean());
                    multiTypeAdapter.notifyDataSetChanged();
                    pageIndex++;
                } else if (pageIndex == 3) {
                    loadMoreWrapper.loadNoMore();
//                        loadMoreWrapper.loadMoreFailure();
                } else {
                    loadMoreWrapper.loadCompleted();
                    int startPosition = multiTypeAdapter.getDataList().size() - 1;
                    multiTypeAdapter.addDatas(startPosition, dataBeanList);
                    multiTypeAdapter.notifyItemRangeInserted(startPosition, dataBeanList.size());
                    pageIndex++;
                }
            }
        }, 1000);
    }

    private void showToast(String toastStr) {
        Toast.makeText(this, toastStr, Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/com/senierr/demo/MainActivity.kt
package com.senierr.demo

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.senierr.adapter.internal.MultiTypeAdapter
import com.senierr.adapter.internal.ViewHolderWrapper
import com.senierr.adapter.internal.ViewTypeLinker
import com.senierr.adapter.support.bean.LoadMoreBean
import com.senierr.demo.databinding.ActivityMainBinding
import com.senierr.demo.entity.Data1
import com.senierr.demo.entity.Data2
import com.senierr.demo.entity.IData
import com.senierr.demo.wrapper.Data1Wrapper
import com.senierr.demo.wrapper.Data2Wrapper
import com.senierr.demo.wrapper.LoadMoreWrapper
import com.senierr.demo.wrapper.StateWrapper
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.lang.IllegalStateException
import java.util.*

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding

    private val multiTypeAdapter = MultiTypeAdapter()

    private val data1Wrapper = Data1Wrapper()
    private val data2Wrapper = Data2Wrapper()

    private var loadMoreWrapper = LoadMoreWrapper()
    private var stateWrapper = StateWrapper()

    private var pageIndex = 0
    private val pageSize = 20

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        initView()
        loadData()
    }

    /**
     * 初始化界面
     */
    private fun initView() {
//        recyclerView.setLayoutManager(new LinearLayoutManager(this));
//        recyclerView.setLayoutManager(new GridLayoutManager(this, 3));
        binding.rvList.layoutManager = StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)

        // 列表点击事件
        data1Wrapper.setOnItemClickListener { _, _, item ->
            showToast("ItemClick: " + item.content)
        }
        data1Wrapper.setOnItemLongClickListener { _, _, item ->
            showToast("ItemLongClick: " + item.content)
            true
        }
        // 子控件点击事件
        data1Wrapper.setOnChildClickListener(R.id.btn_click) { _, _, _, dataBean ->
            showToast("ChildClick: " + dataBean.content)
        }
        data1Wrapper.setOnChildLongClickListener(R.id.btn_click) { _, _, _, dataBean ->
            showToast("ChildLongClick: " + dataBean.content)
            true
        }
        // 加载更多
        loadMoreWrapper.onLoadMoreListener = {
            loadData()
        }
        // 状态切换
        stateWrapper.setOnItemClickListener { _, _, _ ->
            loadData(true)
        }

        multiTypeAdapter.register(
            listOf(data1Wrapper, data2Wrapper, loadMoreWrapper, stateWrapper),
            object : ViewTypeLinker {
                override fun getItemViewType(item: Any): Int {
                    if (item is IData && item.getType() == 1) {
                        return 1
                    } else if (item is IData && item.getType() == 2) {
                        return 2
                    } else if (item is LoadMoreBean) {
                        return 3
                    } else {
                        return 0
                    }
                }

                override fun getViewHolderWrapper(itemViewType: Int): ViewHolderWrapper<*> {
                    return when (itemViewType) {
                        0 -> { stateWrapper }
                        1 -> { data1Wrapper }
                        2 -> { data2Wrapper }
                        3 -> { loadMoreWrapper }
                        else -> {
                            throw IllegalStateException("")
                        }
                    }
                }
            })

        binding.rvList.adapter = multiTypeAdapter
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.action_default) {
            stateWrapper.setState(StateWrapper.STATE_DEFAULT)
        }
        return super.onOptionsItemSelected(item)
    }

    @SuppressLint("NotifyDataSetChanged")
    private fun loadData(isRefresh: Boolean = false) {
        if (isRefresh) pageIndex = 0
        lifecycleScope.launch {
            delay(2000)
            val result = fetchData(pageIndex)
            if (pageIndex == 0) {
                multiTypeAdapter.data.clear()
                multiTypeAdapter.data.addAll(result)
                multiTypeAdapter.data.add(loadMoreWrapper.loadMoreBean)
                multiTypeAdapter.notifyDataSetChanged()
                pageIndex++
            } else if (pageIndex > 3) {
                if ((System.currentTimeMillis() % 2) == 0L) {
                    loadMoreWrapper.loadFailure()
                } else {
                    loadMoreWrapper.loadNoMore()
                }
            } else {
                loadMoreWrapper.loadCompleted()
                val startPosition = multiTypeAdapter.data.size - 1
                multiTypeAdapter.data.addAll(startPosition, result)
                multiTypeAdapter.notifyItemRangeInserted(startPosition, result.size)
                pageIndex++
            }
        }
    }

    private fun fetchData(pageIndex: Int): List<IData> {
        val result1 = mutableListOf<Data1>()
        for (i in 0 until pageSize / 2) {
            result1.add(Data1("$pageIndex-$i", Random().nextInt(300) + 100))
        }
        val result2 = mutableListOf<Data2>()
        for (i in 0 until pageSize / 2) {
            result2.add(Data2("$pageIndex-$i", Random().nextInt(300) + 100))
        }
        return result1 + result2
    }

    private fun showToast(toastStr: String) {
        Toast.makeText(this, toastStr, Toast.LENGTH_SHORT).show()
    }
}
